feat: add subsequent & fuzzy search

Jan-Koll opened the pull request:
Implements an extended and more forgiving search mentioned in #204.

Jan-Koll committed:
Add subsequent & fuzzy search

Jan-Koll committed:
Add preference for fuzzy search, improve search efficiency

Jan-Koll commented on app/src/main/java/de/jrpie/android/launcher/apps/AppFilter.kt:
```diff
@@ -81,6 +104,40 @@ class AppFilter(
+     * A subsequence means all characters in `search` appear in `text`
+     * in the same order, but not necessarily contiguously.
+     */
+    fun isSubsequent(text: String, search: String): Boolean {
```
I agree! Any recommendations/preferences where I should put this exactly? In the existing Functions.kt maybe or in a new file apps/Utils.kt or something like this?

Jan-Koll committed:
Re-enable  check when fuzzy search is enabled

Jan-Koll committed:
Reduce computation when fuzzy search disabled

Respect the fuzzy search preference again to stop computing the fuzzy results when it is not enabled

Jan-Koll committed:
Refactor, improve search accuracy when multiple `contains` results

If multiple items contain the search string perfectly the fuzzy search still expand it with suggestions. These fuzzy-suggestions, however, might have **less** occurrences in the search string than the items that contained it. The changes ensure that fuzzy-suggestions must have at least the same amount of occurrences as apps that contain the search string. Also, I removed the `appsSecondary` wrapper-list and directly work on `r`.

jrpie commented on app/src/main/java/de/jrpie/android/launcher/apps/AppFilter.kt:
```diff
@@ -63,22 +64,71 @@ class AppFilter(
+            val r: MutableSet<AbstractDetailedAppInfo> = hashSetOf()
             val normalizedQuery: String = normalize(query)
+            val subsequentResult: MutableList<AbstractDetailedAppInfo> = mutableListOf();
+            val occurrences: MutableMap<AbstractDetailedAppInfo, Int> = mutableMapOf();
```
I think the approach of having multiple lists breaks down at this point. The approach with two list was mainly used to avoid sorting, but I think the code is getting very hard to read now.
I'd suggest calculating a score for each app, remove all where the score is too high, then sort them by score (and secondarily by name).

Jan-Koll committed:
Remove semicolons

Jan-Koll committed:
Move algorithms to separate file

Jan-Koll committed:
Fix sorting of fuzzy results

Jan-Koll committed:
Improve performance of occurrence counting

Changed the algorithm for occurrence counting to improve performance. Previously it was based on expensive string manipulations and checks. The new approach is based on a frequency map with fewer iterations.

Dynge commented:
Would greatly appreciate this feature.

Any way I can help move it along ?
